clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -229,8 +229,8 @@ private void completeFutureAndFireCallbacks(long baseOffset, long logAppendTime,
                     if (thunk.callback != null)
                         thunk.callback.onCompletion(null, exception);
                 }
-            } catch (Exception e) {
-                log.error("Error executing user-provided callback on message for topic-partition '{}'", topicPartition, e);
+            } catch (Throwable t) {
```

Comment:
It's generally a bad idea to catch throwable.  Why is this needed?

Reply:
Venice team encounters an error that their producer stuck on waiting for memory. The actual error happened several days ago (the producer continues to function after the initial error).  java.lang.OutOfMemoryError is thrown from the user callback, and it is not caught by the exception handlers.  As a result, the batch is not cleaned up, and leave the Kafka producer in an unstable state.  Since Kafka client doesn't have control over the user callback, and the callback is executed in the kafka client thread,  it is ok to isolate user errors from kafka threads.  What do you think?

Reply:
The only safe action to take when OOM happe s is to exit.

Reply:
The best approach is for users to handle these errors in the callback, which Venice team tries to fix.  On our side,  I think catching throwables and not leave kafka client in an unstable state have some benefits.  For example, if users eventually detect the error by other means, they can still gracefully flush pending records and shut down kafka producer.  In addition, some errors may not require close/restart kafka producers.

Reply:
No, they can't gracefully shutdown.

Anything they need to do probably requires memory allocation which is now likely to fail.  We also can't make any guarantees about the behavior of our clients after OOM for the same reasons.

Reply:
The JDK documentation itself says you should not be doing this.  See:

https://docs.oracle.com/javase/7/docs/api/java/lang/Error.html

> An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch. Most such errors are abnormal conditions. The ThreadDeath error, though a "normal" condition, is also a subclass of Error because most applications should not try to catch it.

Reply:
there's no way  to recover from OutOfMem. its not guaranteed that the OOM will be thrown on the same code/thread thats consuming the memory.

the only thing you can do on OOM is die, hence i doubt its worth catching.
worse, catching this will mask the OOM?

Reply:
Ok, the issue here is the error is invisible from users since the callback is executed on the kafka client's thread.  I am ok not to swallow the error (throwable), but we might need a better way to propagate the error to the user so that user can act appropriately.

Reply:
throw it out of the next poll() call ?

Reply:
this is kafka producer.  A simple solution is to catch exception and throwable, log the error if producer see the exceptions, but close the sender thread (unrecoverable) if producer see throwables